tests/test_dimension.py

```diff
@@ -830,6 +831,105 @@ def test_implicit_dims(self):
 
         assert np.all(f.data == F)
 
+    def test_condition_lowering(self):
+        """
+        Test the lowering of a ConditionalDimension's condition.
+        This test makes an Operator that should indexify and lower the condition
+        passed in the Conditional Dimension
+        """
+
+        grid = Grid(shape=(3, 3))  # Define grid
+
+        # Define Function
+        g = Function(name='g', shape=grid.shape, dimensions=grid.dimensions)
+        g.data[:] = 1
+        x, y = grid.dimensions
+
+        ci = ConditionalDimension(name='ci', parent=y, condition=Le(g, 1))
+
+        f = Function(name='f', shape=grid.shape, dimensions=(x, ci))
+        Operator(Eq(f, g)).apply()
+        assert np.all(f.data == 1)
+
+    @pytest.mark.parametrize('setup_rel, rhs, c1, c2, c3, c4', [
+        # Relation, RHS, c1 to c4 used as indexes in assert
+        (Lt, 3, 2, 4, 4, -1), (Le, 2, 2, 4, 4, -1), (Ge, 3, 4, 6, 1, 4),
+        (Gt, 2, 4, 6, 1, 4)
+    ])
+    def test_relational_classes(self, setup_rel, rhs, c1, c2, c3, c4):
+        """
+        Test ConditionalDimension using conditions constructed with overriden
+        SymPy relation classes.
+        """
+
+        class InnerDomain(SubDomain):
+            name = 'inner'
+
+            def define(self, dimensions):
+                return {d: ('middle', 2, 2) for d in dimensions}
+
+        inner_domain = InnerDomain()
+        grid = Grid(shape=(8, 8), subdomains=(inner_domain,))
+        g = Function(name='g', shape=grid.shape, dimensions=grid.dimensions)
+
+        g.data[:4, :4] = 1
+        g.data[4:, :4] = 2
+        g.data[4:, 4:] = 3
+        g.data[:4, 4:] = 4
+
+        x, y = grid.dimensions
+        xi, yi = grid.subdomains['inner'].dimensions
+
+        cond = setup_rel(g[xi, yi], rhs, subdomain=grid.subdomains['inner'])
+        ci = ConditionalDimension(name='ci', parent=yi, condition=cond)
+
+        f = Function(name='f', shape=grid.shape, dimensions=(xi, ci))
+
+        Eq1 = Eq(f[x, y], g[x, y])
+        Eq2 = Eq(f, 5)
+
+        Operator([Eq1, Eq2]).apply()
+
+        assert np.all(f.data[2:6, c1:c2] == 5.)
+        assert np.all(f.data[:, c3:c4] < 5.)
+
+    def test_relational_subdomains(self):
+        """
+        Test a ConditionalDimension applied in a subdomain II.
+        """
+
+        class InnerDomain(SubDomain):
+            name = 'inner'
+
+            def define(self, dimensions):
+                return {d: ('middle', 2, 2) for d in dimensions}
+
+        inner_domain = InnerDomain()
+        grid = Grid(shape=(8, 8), subdomains=(inner_domain,))
+        g = Function(name='g', shape=grid.shape, dimensions=grid.dimensions)
+
+        g.data[:4, :4] = 1
+        g.data[4:, :4] = 2
+        g.data[4:, 4:] = 3
+        g.data[:4, 4:] = 4
+
+        x, y = grid.dimensions
+        xi, yi = grid.subdomains['inner'].dimensions
+
+        cond = And(g[xi, yi] < 3, subdomain=grid.subdomains['inner'])
```

Comment:
This test actually is here to check the generated code fo the unorthodox API code generation through `And`.
I can remove but as discussed this is still a hacky way to semantically have equivalent:
`cond = And(g[xi, yi] < 3, subdomain=grid.subdomains['inner']) `
`cond = Le(g[xi, yi], 3, subdomain=grid.subdomains['inner']) `

Though after this PR, we should drop probably.

Reply:
I'd rather not let people get used to the `And` one and only keep `Le`. I don't think many people use that yet so it shouldn't be an issue to only keep `Le`

Reply:
Agreed. Removed.